### ISSUE TITLE
Feature request: Include browser tabs as if they were windows

### ISSUE THREAD

**Issue by NigelThorne** [tvjg/iswitchw#1]
Thanks for writing this. It's awesome. 

I'd love to be able to select browser tabs from the list too. 
Selecting the tab's title should make the windows active and make that tab active in the browser. 
Then I could go to a specific tab I know I have open somewhere. 

That would be awesome!

**Comment by tvjg** [comment]
Thanks for the kind words. I'm without a Windows machine at the moment, so it might be a bit before I can look into this. 

Off the top of my head, I'm not sure how it would work. Did you have a particular browser in mind? Do you know if any browsers expose an interface to ask for a list of tabs?

**Comment by NigelThorne** [comment]
I think you would have to address each browser independently. I use Chrome
mostly.

We would first need to list all tabs, then when one was selected, give the
window focus and select the tab within the browser.

Here are some links related to selecting tabs in browsers:
https://autohotkey.com/board/topic/74649-how-to-close-a-single-tab-in-a-browser-ff-ie-chrome/
https://autohotkey.com/board/topic/50251-activating-chrome-tab/
https://autohotkey.com/boards/viewtopic.php?t=579

It sounds like the http://www.autohotkey.com/forum/viewtopic.php?t=82767
(ACC library) is the key.

---

"Man, I'm going to have so many chickens when this lot hatch!"

**Comment by mcleodd10** [comment]
Hey there, this is super old, but my fork I just put up implements exactly this. Give it a shot if you're still interested.

https://github.com/evilmanimani/iswitchw

**Comment by void285** [comment]
> Hey there, this is super old, but my fork I just put up implements exactly this. Give it a shot if you're still interested.
> 
> https://github.com/evilmanimani/iswitchw

Hi, could you support tcmatch is your fork, it's usefull when there are many tabs. It's here: https://www.ghisler.ch/board/viewtopic.php?t=22592, and this is a similar project which support tcmatch: https://github.com/goreliu/runz

**Comment by mcleodd10** [comment]
edit: Turns out it was pretty straightforward, and the search capability is definitely more robust :) not fully tested so it's optional, but enabled by default if you want to have a look.

Interesting, I'll definitely take a look, which specific search features of tcmatch do you use, i,e, regex, multiple language support, etc? From what I could gather from that other project, it looks like it's just a simple dllcall, so if that's the case it shouldn't be much trouble.

That other project looks interesting, despite not being able to understand most of it. Looks like a program/shortcut launcher, I suppose it doesn't support browser tabs though?

**Comment by void285** [comment]
It works well, thank you!

Tcmatch support multiple language, for me, I use the PinYin search most. It works like this, I want to search **你好**(which means hello), but too lazy to type the characters, so I just type **nh** to find it. nh is first letter combination of **ni3 hao3**, and ni3 hao3 is the pronounciation tag (PinYin) for 你好. So 你好 -> ni3 hao3 -> nh, I just type two keys to finish the search.

For tcmatch to support PinYin search, I put **tcmatch.ini** and **tcmatch.tbl** in the same folder as tcmatch.dll, which can be 
 found here: https://sl.yunclass.com:201/Necessary/Totalcmd/Plugins/wlx/Quicksearch/, tcmatch.tbl is unicode to pinyin table, with different tcmatch.tbl, it supports other lauguage like korean, or other Chinese character tag systerm like WuBi.

About RunZ, you are right, it is a application launcher with some usefull tools like calculator and qrcode generator, it can't switch applications.

**Comment by mcleodd10** [comment]
Ah neat, so you got the pinyin support working just by adding those files? I actually put in rudimentary (for now) support for adding folders for it to scan that can then be launched along with the tabs, I believe it should attempt to scan your desktop and documents folders to start with. Let me know if you have any other suggestions, cheers.

**Comment by void285** [comment]
Yes it supports pinyin search after I put tcmatch.tbl and tcmatch.ini to the lib folder, it's designed to load the tbl file when available. I know it also supports korean from my tcmatch.ini version in use:
```
[general]
simple_search_activate_char=
regex_search_activate_char=?
leven_search_activate_char=<
srch_activate_char=*
preset_activate_char=>
simple_search_match_beginning_activate_char=^
and_separator_char=&
or_separator_char=|
wdx_separator_char=/
negate_char=!
case_sensitive=0
allow_empty_result=1
filter_files_and_folders=3
match_beginning=0
use_pinyin=1
use_korean=0
[gui]
override_search=1
invert_result=0
[wdx]
debug_output=0
wdx_cache=1000
```

Do you mean you will support program launch in your fork? I know RunZ will scan A_ProgramsCommon and A_StartMenu by default.

PS: I found iswitch miss firefox in its list, not firefox tabs, but firefox itself, maybe it's a version support related issue? My firefox version is 88.0

**Comment by mcleodd10** [comment]
Maybe open a new issue on the fork page for further discussion. However for tabs, I opted to not show the browser process since it would mostly just show a duplicate tab for each open window, unless there's another reason to keep the process separate that I'm not considering? I could make it optional either way

**Comment by void285** [comment]
OK Please open issue tab in your fork.
About firefox, I think it's a version support related issue, my firefox v88.0 tabs don't show in the switch list, and firefox progress itself does not show either, I thought it's a bug, but it's not if this is by design. Everything will be OK when firefox version support be done. I don't use firefox often, I mentioned it just in case it's a ignored problem.
